File: iotdb-client/client-py/flight_sql_example.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

"""
Flight SQL Example: Query IoTDB via ADBC and convert results to pandas DataFrame.

This example demonstrates how to use the ADBC Flight SQL driver to:
1. Connect to IoTDB's Arrow Flight SQL service
2. Execute SQL queries
3. Convert query results directly to pandas DataFrame

Requirements:
    pip install adbc_driver_flightsql pyarrow pandas

Before running, ensure IoTDB is running with Arrow Flight SQL enabled:
    - Set enable_arrow_flight_sql_service=true in iotdb-system.properties
    - Default Flight SQL port is 8904
"""

import adbc_driver_flightsql.dbapi as flight_sql
import pandas as pd

# IoTDB Flight SQL connection parameters
FLIGHT_SQL_URI = "grpc://localhost:8904"
USERNAME = "root"
PASSWORD = "root"


def get_connection():
    """Create an ADBC Flight SQL connection to IoTDB."""
    return flight_sql.connect(
        uri=FLIGHT_SQL_URI,
        db_kwargs={"username": USERNAME, "password": PASSWORD},
    )


def setup_data(conn):
    """Create database, table, and insert sample data."""
    cursor = conn.cursor()

    # Create database
    cursor.execute("CREATE DATABASE IF NOT EXISTS flight_example_db")

    # Create table
    cursor.execute(
        "CREATE TABLE IF NOT EXISTS flight_example_db.sensor_data ("
        "region STRING TAG, "
        "device_id STRING TAG, "
        "temperature FLOAT FIELD, "
        "humidity DOUBLE FIELD, "
        "status BOOLEAN FIELD, "
        "description TEXT FIELD)"
    )

    # Insert sample data
    insert_sqls = [
        "INSERT INTO flight_example_db.sensor_data"
        "(time, region, device_id, temperature, humidity, status, description) "
        "VALUES(1, 'north', 'dev001', 25.5, 60.2, true, 'normal')",
        "INSERT INTO flight_example_db.sensor_data"
        "(time, region, device_id, temperature, humidity, status, description) "
        "VALUES(2, 'north', 'dev001', 26.1, 58.5, true, 'normal')",
        "INSERT INTO flight_example_db.sensor_data"
        "(time, region, device_id, temperature, humidity, status, description) "
        "VALUES(3, 'south', 'dev002', 30.2, 45.0, false, 'high temp')",
        "INSERT INTO flight_example_db.sensor_data"
        "(time, region, device_id, temperature, humidity, status, description) "
        "VALUES(4, 'south', 'dev002', 31.5, 42.3, false, 'high temp')",
        "INSERT INTO flight_example_db.sensor_data"
        "(time, region, device_id, temperature, humidity, status, description) "
        "VALUES(5, 'north', 'dev003', 22.0, 70.1, true, 'cool')",
    ]
    for sql in insert_sqls:
        cursor.execute(sql)

    cursor.close()
    print("Sample data inserted successfully.")


def query_to_dataframe(conn, sql):
    """Execute a SQL query and return results as a pandas DataFrame."""
    cursor = conn.cursor()
    cursor.execute(sql)
    # Fetch result as Arrow Table, then convert to pandas DataFrame
    arrow_table = cursor.fetch_arrow_table()
    df = arrow_table.to_pandas()
    cursor.close()
    return df


def main():
    print("=" * 60)
    print("IoTDB Arrow Flight SQL Example (Python ADBC)")
    print("=" * 60)

    # Connect to IoTDB Flight SQL service
    conn = get_connection()
    print(f"\nConnected to IoTDB Flight SQL at {FLIGHT_SQL_URI}")

    # Setup sample data
    print("\n--- Setting up test data ---")
    setup_data(conn)

    # Example 1: Full table scan → DataFrame
    print("\n" + "=" * 60)
    print("Example 1: Full table scan → DataFrame")
    print("=" * 60)
    df = query_to_dataframe(
        conn,
        "SELECT time, region, device_id, temperature, humidity, status, description "
        "FROM flight_example_db.sensor_data ORDER BY time",
    )
    print(f"\nDataFrame shape: {df.shape}")
    print(f"Column types:\n{df.dtypes}\n")
    print(df.to_string(index=False))

    # Example 2: Filtered query → DataFrame
    print("\n" + "=" * 60)
    print("Example 2: Filtered query (region = 'north') → DataFrame")
    print("=" * 60)
    df_north = query_to_dataframe(
        conn,
        "SELECT time, device_id, temperature, humidity "
        "FROM flight_example_db.sensor_data WHERE region = 'north' ORDER BY time",
    )
    print(f"\nDataFrame shape: {df_north.shape}")
    print(df_north.to_string(index=False))

    # Example 3: Aggregation query → DataFrame
    print("\n" + "=" * 60)
    print("Example 3: Aggregation → DataFrame")
    print("=" * 60)
    df_agg = query_to_dataframe(
        conn,
        "SELECT region, COUNT(*) as cnt, AVG(temperature) as avg_temp, "
        "MAX(humidity) as max_humidity "
        "FROM flight_example_db.sensor_data GROUP BY region ORDER BY region",
    )
    print(f"\nAggregation DataFrame shape: {df_agg.shape}")
    print(df_agg.to_string(index=False))

    # Example 4: Using pandas operations on the DataFrame
    print("\n" + "=" * 60)
    print("Example 4: Pandas operations on Flight SQL results")
    print("=" * 60)
    df_all = query_to_dataframe(
        conn,
        "SELECT time, region, device_id, temperature, humidity "
        "FROM flight_example_db.sensor_data ORDER BY time",
    )
    # Compute statistics
    print(f"\nTemperature statistics:\n{df_all['temperature'].describe()}")
    print(f"\nGroup by region mean:\n{df_all.groupby('region')[['temperature', 'humidity']].mean()}")

    # Cleanup
    cursor = conn.cursor()
    cursor.execute("DROP DATABASE IF EXISTS flight_example_db")
    cursor.close()

    conn.close()
    print("\n" + "=" * 60)
    print("Example completed successfully!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: example/flight-sql-example/src/main/java/org/apache/iotdb/FlightSqlExample.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.iotdb;

import org.apache.arrow.flight.FlightClient;
import org.apache.arrow.flight.FlightEndpoint;
import org.apache.arrow.flight.FlightInfo;
import org.apache.arrow.flight.FlightStream;
import org.apache.arrow.flight.Location;
import org.apache.arrow.flight.auth2.BasicAuthCredentialWriter;
import org.apache.arrow.flight.auth2.ClientBearerHeaderHandler;
import org.apache.arrow.flight.auth2.ClientIncomingAuthHeaderMiddleware;
import org.apache.arrow.flight.grpc.CredentialCallOption;
import org.apache.arrow.flight.sql.FlightSqlClient;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.FieldVector;
import org.apache.arrow.vector.VectorSchemaRoot;
import org.apache.arrow.vector.types.pojo.Field;
import org.apache.arrow.vector.types.pojo.Schema;

import java.util.List;

/**
 * Example demonstrating how to use Arrow Flight SQL to query IoTDB.
 *
 * <p>Before running this example, make sure IoTDB is running with Arrow Flight SQL enabled:
 *
 * <ul>
 *   <li>Set <code>enable_arrow_flight_sql_service=true</code> in iotdb-system.properties
 *   <li>Default Flight SQL port is 8904
 * </ul>
 */
public class FlightSqlExample {

  private static final String HOST = "127.0.0.1";
  private static final int PORT = 8904;
  private static final String USER = "root";
  private static final String PASSWORD = "root";

  public static void main(String[] args) throws Exception {
    // 1. Create Arrow allocator and Flight SQL client
    try (BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE)) {
      Location location = Location.forGrpcInsecure(HOST, PORT);

      // Set up Bearer token authentication middleware
      ClientIncomingAuthHeaderMiddleware.Factory authFactory =
          new ClientIncomingAuthHeaderMiddleware.Factory(new ClientBearerHeaderHandler());

      try (FlightClient flightClient =
              FlightClient.builder(allocator, location).intercept(authFactory).build();
          FlightSqlClient sqlClient = new FlightSqlClient(flightClient)) {

        // 2. Authenticate with username/password to get Bearer token
        CredentialCallOption bearerToken =
            new CredentialCallOption(new BasicAuthCredentialWriter(USER, PASSWORD));
        System.out.println("=== Connected to IoTDB Flight SQL service ===");
        System.out.println("Host: " + HOST + ", Port: " + PORT);

        // 3. Create database and table, insert data
        System.out.println("\n--- Setting up test data ---");
        executeUpdate(sqlClient, "CREATE DATABASE IF NOT EXISTS flight_example_db", bearerToken);
        executeUpdate(
            sqlClient,
            "CREATE TABLE IF NOT EXISTS flight_example_db.sensor_data ("
                + "region STRING TAG, "
                + "device_id STRING TAG, "
                + "temperature FLOAT FIELD, "
                + "humidity DOUBLE FIELD, "
                + "status BOOLEAN FIELD, "
                + "description TEXT FIELD)",
            bearerToken);

        // Insert sample data
        executeUpdate(
            sqlClient,
            "INSERT INTO flight_example_db.sensor_data"
                + "(time, region, device_id, temperature, humidity, status, description) "
                + "VALUES(1, 'north', 'dev001', 25.5, 60.2, true, 'normal')",
            bearerToken);
        executeUpdate(
            sqlClient,
            "INSERT INTO flight_example_db.sensor_data"
                + "(time, region, device_id, temperature, humidity, status, description) "
                + "VALUES(2, 'north', 'dev001', 26.1, 58.5, true, 'normal')",
            bearerToken);
        executeUpdate(
            sqlClient,
            "INSERT INTO flight_example_db.sensor_data"
                + "(time, region, device_id, temperature, humidity, status, description) "
                + "VALUES(3, 'south', 'dev002', 30.2, 45.0, false, 'high temp')",
            bearerToken);
        executeUpdate(
            sqlClient,
            "INSERT INTO flight_example_db.sensor_data"
                + "(time, region, device_id, temperature, humidity, status, description) "
                + "VALUES(4, 'south', 'dev002', 31.5, 42.3, false, 'high temp')",
            bearerToken);
        System.out.println("Test data inserted successfully.");

        // 4. Query: SELECT all data
        System.out.println("\n=== Query 1: Full table scan ===");
        executeQuery(
            sqlClient,
            "SELECT time, region, device_id, temperature, humidity, status, description "
                + "FROM flight_example_db.sensor_data ORDER BY time",
            bearerToken);

        // 5. Query with filter
        System.out.println("\n=== Query 2: Filtered query (region = 'north') ===");
        executeQuery(
            sqlClient,
            "SELECT time, device_id, temperature, humidity "
                + "FROM flight_example_db.sensor_data WHERE region = 'north' ORDER BY time",
            bearerToken);

        // 6. Aggregation query
        System.out.println("\n=== Query 3: Aggregation (AVG temperature by region) ===");
        executeQuery(
            sqlClient,
            "SELECT region, COUNT(*) as cnt, AVG(temperature) as avg_temp "
                + "FROM flight_example_db.sensor_data GROUP BY region ORDER BY region",
            bearerToken);

        // 7. Cleanup
        executeUpdate(sqlClient, "DROP DATABASE IF EXISTS flight_example_db", bearerToken);
        System.out.println("\n=== Example completed successfully ===");
      }
    }
  }

  /** Execute a non-query SQL statement (DDL/DML). */
  private static void executeUpdate(
      FlightSqlClient sqlClient, String sql, CredentialCallOption bearerToken) throws Exception {
    FlightInfo info = sqlClient.execute(sql, bearerToken);
    // Consume the stream to ensure execution
    for (FlightEndpoint endpoint : info.getEndpoints()) {
      try (FlightStream stream = sqlClient.getStream(endpoint.getTicket(), bearerToken)) {
        while (stream.next()) {
          // consume
        }
      }
    }
  }

  /** Execute a query and print results in table format. */
  private static void executeQuery(
      FlightSqlClient sqlClient, String sql, CredentialCallOption bearerToken) throws Exception {
    FlightInfo flightInfo = sqlClient.execute(sql, bearerToken);

    // Print schema
    Schema schema = flightInfo.getSchema();
    List<Field> fields = schema.getFields();
    System.out.println("SQL: " + sql);
    System.out.println("Schema: " + fields.size() + " columns");

    // Print column headers
    StringBuilder header = new StringBuilder();
    StringBuilder separator = new StringBuilder();
    for (Field field : fields) {
      String colName = field.getName();
      header.append(String.format("%-20s", colName));
      separator.append("--------------------");
    }
    System.out.println(header);
    System.out.println(separator);

    // Fetch and print rows
    int totalRows = 0;
    for (FlightEndpoint endpoint : flightInfo.getEndpoints()) {
      try (FlightStream stream = sqlClient.getStream(endpoint.getTicket(), bearerToken)) {
        while (stream.next()) {
          VectorSchemaRoot root = stream.getRoot();
          int rowCount = root.getRowCount();
          for (int i = 0; i < rowCount; i++) {
            StringBuilder row = new StringBuilder();
            for (FieldVector vector : root.getFieldVectors()) {
              Object value = vector.getObject(i);
              row.append(String.format("%-20s", value == null ? "null" : value.toString()));
            }
            System.out.println(row);
            totalRows++;
          }
        }
      }
    }
    System.out.println("Total rows: " + totalRows);
  }
}
